File: levanta/android/core/src/main/kotlin/br/com/levanta/core/pose/Exercicios.kt
package br.com.levanta.core.pose

import br.com.levanta.core.pose.Marco.COTOVELO_DIR
import br.com.levanta.core.pose.Marco.COTOVELO_ESQ
import br.com.levanta.core.pose.Marco.JOELHO_DIR
import br.com.levanta.core.pose.Marco.JOELHO_ESQ
import br.com.levanta.core.pose.Marco.OMBRO_DIR
import br.com.levanta.core.pose.Marco.OMBRO_ESQ
import br.com.levanta.core.pose.Marco.PULSO_DIR
import br.com.levanta.core.pose.Marco.PULSO_ESQ
import br.com.levanta.core.pose.Marco.QUADRIL_DIR
import br.com.levanta.core.pose.Marco.QUADRIL_ESQ
import br.com.levanta.core.pose.Marco.TORNOZELO_DIR
import br.com.levanta.core.pose.Marco.TORNOZELO_ESQ

/** Fábrica: o contador certo para o exercício escolhido. */
fun contadorDe(exercicio: Exercicio): ContadorRepeticoes = when (exercicio) {
    Exercicio.AGACHAMENTO -> ContadorAgachamento()
    Exercicio.FLEXAO -> ContadorFlexao()
    Exercicio.POLICHINELO -> ContadorPolichinelo()
    Exercicio.ABDOMINAL -> ContadorAbdominal()
}

/** Eixo do tronco: do meio dos ombros ao meio do quadril. */
private fun tronco(e: Esqueleto): Pair<Ponto, Ponto>? {
    val ombro = medio(e.ponto(OMBRO_ESQ), e.ponto(OMBRO_DIR)) ?: return null
    val quadril = medio(e.ponto(QUADRIL_ESQ), e.ponto(QUADRIL_DIR)) ?: return null
    return ombro to quadril
}

/**
 * Agachamento — ângulo do joelho (quadril · joelho · tornozelo).
 *
 * Em pé passa de 170°; na coxa paralela ao chão fica perto de 80°. Os limiares
 * ficam em 160° e 100°, então o agachamento tem que ser fundo de verdade: parar
 * a meio caminho, por volta de 130°, não cruza nenhum dos dois e não conta.
 */
class ContadorAgachamento : ContadorRepeticoes(Exercicio.AGACHAMENTO) {
    override val limiarRepouso = 160f
    override val limiarExtremo = 100f

    override fun sinal(e: Esqueleto) = melhorLado(e) { esq, lado ->
        when (lado) {
            Lado.ESQUERDO -> angulo(esq.ponto(QUADRIL_ESQ), esq.ponto(JOELHO_ESQ), esq.ponto(TORNOZELO_ESQ))
            Lado.DIREITO -> angulo(esq.ponto(QUADRIL_DIR), esq.ponto(JOELHO_DIR), esq.ponto(TORNOZELO_DIR))
        }
    }

    /** De pé: o tronco tem que estar mais vertical que horizontal. */
    override fun postura(e: Esqueleto): Aviso? {
        val (ombro, quadril) = tronco(e) ?: return Aviso.CORPO_FORA_DE_QUADRO
        val inc = inclinacao(ombro, quadril) ?: return Aviso.CORPO_FORA_DE_QUADRO
        return if (inc < 40f) Aviso.POSICAO_INVALIDA else null
    }
}

/**
 * Flexão de braço — ângulo do cotovelo (ombro · cotovelo · pulso).
 *
 * Braço estendido fica acima de 165°; peito perto do chão, abaixo de 90°.
 */
class ContadorFlexao : ContadorRepeticoes(Exercicio.FLEXAO) {
    override val limiarRepouso = 155f
    override val limiarExtremo = 100f

    override fun sinal(e: Esqueleto) = melhorLado(e) { esq, lado ->
        when (lado) {
            Lado.ESQUERDO -> angulo(esq.ponto(OMBRO_ESQ), esq.ponto(COTOVELO_ESQ), esq.ponto(PULSO_ESQ))
            Lado.DIREITO -> angulo(esq.ponto(OMBRO_DIR), esq.ponto(COTOVELO_DIR), esq.ponto(PULSO_DIR))
        }
    }

    /**
     * Prancha: tronco quase horizontal e quadril alinhado com ombro e joelho.
     * Sem isso, dobrar o cotovelo sentado na cama contaria como flexão.
     */
    override fun postura(e: Esqueleto): Aviso? {
        val (ombro, quadril) = tronco(e) ?: return Aviso.CORPO_FORA_DE_QUADRO
        val inc = inclinacao(ombro, quadril) ?: return Aviso.CORPO_FORA_DE_QUADRO
        if (inc > 40f) return Aviso.POSICAO_INVALIDA
        val joelho = medio(e.ponto(JOELHO_ESQ), e.ponto(JOELHO_DIR))
        val corpoReto = angulo(ombro, quadril, joelho)
        return if (corpoReto != null && corpoReto < 140f) Aviso.POSICAO_INVALIDA else null
    }
}

/**
 * Abdominal — ângulo do quadril (ombro · quadril · joelho).
 *
 * Deitado passa de 130°; com o tronco em cima, perto dos joelhos, cai abaixo
 * de 70°.
 */
class ContadorAbdominal : ContadorRepeticoes(Exercicio.ABDOMINAL) {
    override val limiarRepouso = 130f
    override val limiarExtremo = 80f

    override fun sinal(e: Esqueleto) = melhorLado(e) { esq, lado ->
        when (lado) {
            Lado.ESQUERDO -> angulo(esq.ponto(OMBRO_ESQ), esq.ponto(QUADRIL_ESQ), esq.ponto(JOELHO_ESQ))
            Lado.DIREITO -> angulo(esq.ponto(OMBRO_DIR), esq.ponto(QUADRIL_DIR), esq.ponto(JOELHO_DIR))
        }
    }

    /**
     * No chão: as pernas ficam deitadas o movimento inteiro, então a linha
     * quadril→tornozelo é o que distingue um abdominal de alguém sentado se
     * inclinando para a frente — nesse caso a mesma leitura de quadril
     * apareceria com as pernas na vertical.
     */
    override fun postura(e: Esqueleto): Aviso? {
        val quadril = medio(e.ponto(QUADRIL_ESQ), e.ponto(QUADRIL_DIR)) ?: return Aviso.CORPO_FORA_DE_QUADRO
        val tornozelo = medio(e.ponto(TORNOZELO_ESQ), e.ponto(TORNOZELO_DIR)) ?: return Aviso.CORPO_FORA_DE_QUADRO
        val inc = inclinacao(quadril, tornozelo) ?: return Aviso.CORPO_FORA_DE_QUADRO
        return if (inc > 45f) Aviso.POSICAO_INVALIDA else null
    }
}

/**
 * Polichinelo — índice de fechamento de 0 a 100, alto com o corpo fechado.
 *
 * Não existe um ângulo único que descreva o movimento, então o sinal combina
 * duas medidas em partes iguais: quanto os braços subiram e quanto os pés se
 * afastaram. Cobrar as duas juntas é o que impede contar quem só bate palma em
 * cima ou só arrasta os pés.
 *
 * As duas medidas são divididas pela altura do tronco, e não por pixels ou pela
 * largura dos ombros: assim o mesmo movimento conta igual perto ou longe da
 * câmera, e a altura do tronco continua estável mesmo quando a pessoa gira um
 * pouco de lado, o que encolheria a largura dos ombros.
 */
class ContadorPolichinelo : ContadorRepeticoes(Exercicio.POLICHINELO) {
    override val limiarRepouso = 72f
    override val limiarExtremo = 30f
    override val suavizacao = 0.6f
    override val msNoExtremo = 120L
    override val msPorRepeticao = 400L

    override fun sinal(e: Esqueleto): Float? {
        val (ombro, quadril) = tronco(e) ?: return null
        val alturaTronco = distancia(ombro, quadril) ?: return null
        if (alturaTronco < 1e-4f) return null

        val pulso = medio(e.ponto(PULSO_ESQ), e.ponto(PULSO_DIR)) ?: return null
        val tornozeloEsq = e.ponto(TORNOZELO_ESQ) ?: return null
        val tornozeloDir = e.ponto(TORNOZELO_DIR) ?: return null

        // Braço no quadril dá ~0; acima da cabeça, ~2 alturas de tronco.
        val bracos = faixa((quadril.y - pulso.y) / alturaTronco, 0.25f, 1.5f)
        // Pés juntos dão ~0,2 de tronco entre eles; abertos, ~1,0.
        val pernas = faixa((distancia(tornozeloEsq, tornozeloDir) ?: return null) / alturaTronco, 0.35f, 0.95f)

        return 100f * (1f - (bracos + pernas) / 2f)
    }

    /** Em pé, de frente para a câmera. */
    override fun postura(e: Esqueleto): Aviso? {
        val (ombro, quadril) = tronco(e) ?: return Aviso.CORPO_FORA_DE_QUADRO
        val inc = inclinacao(ombro, quadril) ?: return Aviso.CORPO_FORA_DE_QUADRO
        return if (inc < 55f) Aviso.POSICAO_INVALIDA else null
    }
}

File: levanta/ios/Levanta/Nucleo/Exercicios.swift
import Foundation

/// Eixo do tronco: do meio dos ombros ao meio do quadril.
private func tronco(_ e: Esqueleto) -> (ombro: Ponto, quadril: Ponto)? {
    guard let ombro = medio(e.ponto(.ombroEsq), e.ponto(.ombroDir)),
          let quadril = medio(e.ponto(.quadrilEsq), e.ponto(.quadrilDir)) else { return nil }
    return (ombro, quadril)
}

/// Agachamento — ângulo do joelho (quadril · joelho · tornozelo).
///
/// Em pé passa de 170°; na coxa paralela ao chão fica perto de 80°. Os limiares
/// em 160° e 100° exigem um agachamento fundo: parar a meio caminho, por volta
/// de 130°, não cruza nenhum dos dois e não conta.
final class ContadorAgachamento: ContadorRepeticoes {
    init() { super.init(exercicio: .agachamento) }

    override var limiarRepouso: Float { 160 }
    override var limiarExtremo: Float { 100 }

    override func sinal(_ e: Esqueleto) -> Float? {
        melhorLado(e) { esq, lado in
            switch lado {
            case .esquerdo: angulo(esq.ponto(.quadrilEsq), esq.ponto(.joelhoEsq), esq.ponto(.tornozeloEsq))
            case .direito: angulo(esq.ponto(.quadrilDir), esq.ponto(.joelhoDir), esq.ponto(.tornozeloDir))
            }
        }
    }

    /// De pé: o tronco tem que estar mais vertical que horizontal.
    override func postura(_ e: Esqueleto) -> Aviso? {
        guard let t = tronco(e), let inc = inclinacao(t.ombro, t.quadril) else { return .corpoForaDeQuadro }
        return inc < 40 ? .posicaoInvalida : nil
    }
}

/// Flexão de braço — ângulo do cotovelo (ombro · cotovelo · pulso).
final class ContadorFlexao: ContadorRepeticoes {
    init() { super.init(exercicio: .flexao) }

    override var limiarRepouso: Float { 155 }
    override var limiarExtremo: Float { 100 }

    override func sinal(_ e: Esqueleto) -> Float? {
        melhorLado(e) { esq, lado in
            switch lado {
            case .esquerdo: angulo(esq.ponto(.ombroEsq), esq.ponto(.cotoveloEsq), esq.ponto(.pulsoEsq))
            case .direito: angulo(esq.ponto(.ombroDir), esq.ponto(.cotoveloDir), esq.ponto(.pulsoDir))
            }
        }
    }

    /// Prancha: tronco quase horizontal e quadril alinhado com ombro e joelho.
    /// Sem isso, dobrar o cotovelo sentado na cama contaria como flexão.
    override func postura(_ e: Esqueleto) -> Aviso? {
        guard let t = tronco(e), let inc = inclinacao(t.ombro, t.quadril) else { return .corpoForaDeQuadro }
        if inc > 40 { return .posicaoInvalida }
        let joelho = medio(e.ponto(.joelhoEsq), e.ponto(.joelhoDir))
        if let corpoReto = angulo(t.ombro, t.quadril, joelho), corpoReto < 140 { return .posicaoInvalida }
        return nil
    }
}

/// Abdominal — ângulo do quadril (ombro · quadril · joelho).
final class ContadorAbdominal: ContadorRepeticoes {
    init() { super.init(exercicio: .abdominal) }

    override var limiarRepouso: Float { 130 }
    override var limiarExtremo: Float { 80 }

    override func sinal(_ e: Esqueleto) -> Float? {
        melhorLado(e) { esq, lado in
            switch lado {
            case .esquerdo: angulo(esq.ponto(.ombroEsq), esq.ponto(.quadrilEsq), esq.ponto(.joelhoEsq))
            case .direito: angulo(esq.ponto(.ombroDir), esq.ponto(.quadrilDir), esq.ponto(.joelhoDir))
            }
        }
    }

    /// No chão: as pernas ficam deitadas o movimento inteiro. É o que distingue
    /// um abdominal de alguém sentado se inclinando para a frente, que produz a
    /// mesma leitura de quadril com as pernas na vertical.
    override func postura(_ e: Esqueleto) -> Aviso? {
        guard let quadril = medio(e.ponto(.quadrilEsq), e.ponto(.quadrilDir)),
              let tornozelo = medio(e.ponto(.tornozeloEsq), e.ponto(.tornozeloDir)),
              let inc = inclinacao(quadril, tornozelo) else { return .corpoForaDeQuadro }
        return inc > 45 ? .posicaoInvalida : nil
    }
}

/// Polichinelo — índice de fechamento de 0 a 100, alto com o corpo fechado.
///
/// Não existe um ângulo único que descreva o movimento, então o sinal combina
/// em partes iguais quanto os braços subiram e quanto os pés se afastaram.
/// Cobrar as duas coisas juntas impede contar quem só bate palma em cima ou só
/// arrasta os pés. As duas medidas são divididas pela altura do tronco: assim o
/// movimento conta igual perto ou longe da câmera, e a referência continua
/// estável mesmo quando a pessoa gira um pouco de lado.
final class ContadorPolichinelo: ContadorRepeticoes {
    init() { super.init(exercicio: .polichinelo) }

    override var limiarRepouso: Float { 72 }
    override var limiarExtremo: Float { 30 }
    override var suavizacao: Float { 0.6 }
    override var msNoExtremo: Int64 { 120 }
    override var msPorRepeticao: Int64 { 400 }

    override func sinal(_ e: Esqueleto) -> Float? {
        guard let t = tronco(e),
              let alturaTronco = distancia(t.ombro, t.quadril), alturaTronco > 1e-4,
              let pulso = medio(e.ponto(.pulsoEsq), e.ponto(.pulsoDir)),
              let tornozeloEsq = e.ponto(.tornozeloEsq),
              let tornozeloDir = e.ponto(.tornozeloDir),
              let separacao = distancia(tornozeloEsq, tornozeloDir) else { return nil }

        // Braço no quadril dá ~0; acima da cabeça, ~2 alturas de tronco.
        let bracos = faixa((t.quadril.y - pulso.y) / alturaTronco, 0.25, 1.5)
        // Pés juntos dão ~0,2 de tronco entre eles; abertos, ~1,0.
        let pernas = faixa(separacao / alturaTronco, 0.35, 0.95)

        return 100 * (1 - (bracos + pernas) / 2)
    }

    /// Em pé, de frente para a câmera.
    override func postura(_ e: Esqueleto) -> Aviso? {
        guard let t = tronco(e), let inc = inclinacao(t.ombro, t.quadril) else { return .corpoForaDeQuadro }
        return inc < 55 ? .posicaoInvalida : nil
    }
}
